rubyforgood/abalone
# #714 : exit types

Sorry for my english.

# Checklist:

- [x] New and existing unit tests pass locally with my changes ("bundle exec rake"),
- [ ] Title include "WIP" if work is in progress.

Resolves #714 

### Description
First the diagram :
![exit_type_schema](https://user-images.githubusercontent.com/84066080/136087967-f4d7f264-8aea-4834-83f4-4bd453e787a9.png)

What can we do now ?
- Go the the exit types index of the current organization with the admin menu
- Add an exit type
- Edit it
- Show it
- Destroy an output type if there is no mortality_events that uses it as a foreign key.
- Validation of the exit type : name is required
- if the user try to show an exit type of another organization then a message is displayed 

### Type of change


* Breaking change (fix or feature that would cause existing functionality to not 
work as expected)

### How Has This Been Tested?
`bundle exec rake`

Use cases of exit type :
- include organization presence validation
- exit_type must have a name

Use cases of exit type controller :
- index should have response code 200 for admin user
- index should have response code 302 for non-admin user
- new should have response code 200 for admin user
- new should have response code 302 for non-admin user
- show should have response code 200 for admin user
- show should have response code 302 for non-admin user
- show should have response code 302 for an exit types in another organization
- edit should have response code 200 for admin user
- edit should have response code 302 for non-admin user
- edit should have response code 302 for an exit types in another organization
- destroy should not destroy if the user is non-admin
- destroy should not destroy for an exit types in another organization
- destroy should not destroy if the exit types referenced in a mortality event
- should have delete the entity for admin user

Use cases of features/create :
- When I visit the exit type New page and fill out the form and click the submit button, exit type should be created
- When I visit the exit type New page and click the submit button without fill the name, an error message must be shown

Use cases of features/destroy :
- When I visit the exit type Index page then I click the delete button, exit type should be deleted

Use cases of features/index :
- When I visit the exit type index page then I see a list of exit_types
- When I visit the exit type index page then I can't see exit_types of another organization
- As a non-admin user then I should not have access to the exit types index page

Use cases of features/show :
- When I visit the exit type Show page I see information of a specific exit type
- When I visit the exit type Show page I can't see an exit type of another organization

Use cases of features/update :
- When I visit the exit type Edit page and fill out the form and click the submit button, exit_type should be updated
- When I visit the exit type Edit page and click the submit button with an empty name, an error message must be shown

### Screenshots
![s1](https://user-images.githubusercontent.com/84066080/136091195-6a19d6eb-99b8-434c-be4a-c72ab95aabee.png)
![menu_link](https://user-images.githubusercontent.com/84066080/136091214-d32fc1d0-8dbd-40fe-9c29-80453d750578.png)
![s2](https://user-images.githubusercontent.com/84066080/136091237-6ea32b1b-2d80-4c65-90af-450b1ef282f7.png)
![s3](https://user-images.githubusercontent.com/84066080/136091246-fb26af97-d4f0-4f77-b7ec-a511f24f8d4d.png)
